fix: promote re-shares when deleting the parent share

come-nc opened the pull request:
* Resolves: # <!-- related github issue -->

## Summary

Same as #43025 but without the new command.

* ~Catch share exception in transfer ownership and delete invalid shares~
* Promote reshares into direct shares when a user loses access to a file after an unshare

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

luka-nextcloud committed:
fix: delete re-shares when deleting the parent share

Note: Removed part about fix command from original PR

Signed-off-by: Luka Trovic <luka@nextcloud.com>
Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

solracsf commented on lib/private/Share20/Manager.php:
```diff
@@ -1017,6 +1018,71 @@ protected function deleteChildren(IShare $share) {
+
+		$userIds = [];
+
+		if  ($share->getShareType() === IShare::TYPE_USER) {
```
```suggestion
		if ($share->getShareType() === IShare::TYPE_USER) {
```

come-nc committed:
fix: Tidy up code for reshare deletion

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc commented:
The failing tests are related.

The test failing is testing this scenario:

> 	Scenario: transferring ownership of folder reshared with another user
> 		Given user "user0" exists
> 		And user "user1" exists
> 		And user "user2" exists
> 		And user "user3" exists
> 		And User "user3" created a folder "/test"
> 		And User "user3" uploads file "data/textfile.txt" to "/test/somefile.txt"
> 		And folder "/test" of user "user3" is shared with user "user0" with permissions 31
> 		And user "user0" accepts last share
> 		And folder "/test" of user "user0" is shared with user "user2" with permissions 31
> 		And user "user2" accepts last share
> 		When transferring ownership from "user0" to "user1"
> 		And the command was successful
> 		And As an "user2"
> 		**Then Downloaded content when downloading file "/test/somefile.txt" with range "bytes=0-6" should be "This is"**
> 		And using old dav path
> 		And as "user0" the folder "/test" exists
> 		And using received transfer folder of "user1" as dav path
> 		And as "user1" the folder "/test" does not exist
> 		And As an "user0"
> 		And Getting info of last share
> 		And the OCS status code should be "100"
> 		And Share fields of last share match with
> 			| uid_owner | user0 |
> 			| uid_file_owner | user3 |
> 			| share_with | user2 |

It fails at the step in bold.
The problem here is that we ask the command to transfer a reshare from user0 to user1 without transferring the incoming share. That would result in a situation where user1 shares the folder without having access to it. The test actually expects it to fail, it checks that user1 does not see the folder.
But with the new code, the share is seen as invalid and is deleted. While the old code would let it in place.

If I reorder the transferring code and use ` --transfer-incoming-shares=1` then I can successfully transfer the share.

The more I think about it the more it feels like it’s not the job of transfer-ownership to delete broken shares.
The fix in the share manager is enough to avoid share breakage on share deletion.
What would remain to fix is to behave better when encountering an already broken share.
From original bug report I see 2 occurences listed:
https://github.com/nextcloud/server/pull/43025#issuecomment-2189118746
https://github.com/nextcloud/server/pull/43025#issuecomment-1914759785

come-nc committed:
fix: Transfer incomming shares first, do not delete non-migratable ones

Canceling the previous add of deletion of invalid shares in
 transferownership because in some cases it deletes valid reshares, if
 incoming shares are not transfered on purpose.
Inverting the order of transfer between incoming and outgoing so that
 reshare can be migrated when incoming shares are transfered.

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc commented:
After internal discussion, reshares should **not** be deleted when share is deleted but turned into direct shares instead.
So something more along the line of what the repair share proposition from Luka did.

We should still be careful about what that means for transfer-ownership when incomming shares are not transfered.

come-nc committed:
fix(shares): Promote reshares into direct shares when share is deleted

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

icewind1991 commented on lib/private/Share20/Manager.php:
```diff
@@ -1010,6 +1012,79 @@ protected function deleteChildren(IShare $share) {
+			} catch (GenericShareException $e) {
+				$this->logger->debug('Promote reshare because of exception '.$e->getMessage(), ['exception' => $e, 'fullId' => $child->getFullId()]);
+				try {
+					$child->setSharedBy($share->getSharedBy());
```
I'm having problems understanding the flow here, why is the actual update done in the `catch` block?

come-nc replied:
Because if the exception does not fire it means the sharer still has access to the file through another way (another group share, share of a parent folder, groupfolder, you name it) and there is no need to update the reshare.

icewind1991 replied:
Ok, I didn't catch that the check was done for the original reshare, not the updated one.

Some comment about this logic would be appreciated

susnux commented on lib/private/Share20/Manager.php:
```diff
@@ -1010,6 +1012,79 @@ protected function deleteChildren(IShare $share) {
 		return $deletedShares;
 	}
 
+	/* Promote reshares into direct shares so that target user keeps access */
```
```suggestion
	/** Promote re-shares into direct shares so that target user keeps access */
```

susnux commented:
To make sure I understand this correct:
1. A shares F with B
2. B re-shares  F with C
3. A deletes share of F with B
4. B -> C share of F is converted to a share A -> C of F?

I think there was a discussion recently with @jancborchardt if this is expected behavior.
If I remember correctly if the original sharer now has control over the shares this should be fine, but C should not have access to F after B is removed without that A can control this.

come-nc commented:
> To make sure I understand this correct:
> 
> 1. A shares F with B
> 2. B re-shares  F with C
> 3. A deletes share of F with B
> 4. B -> C share of F is converted to a share A -> C of F?

Yes that is correct

> I think there was a discussion recently with @jancborchardt if this is expected behavior. If I remember correctly if the original sharer now has control over the shares this should be fine, but C should not have access to F after B is removed without that A can control this.

Before deleting the share to B, A sees shares both to B and C in the UI. There is no visible distinction between the share and the reshare.
After deleting the share to B, the share to C is untouched in the UI, so A knows that he need to delete it if needed. As it’s now a direct share he can also edit the permissions on it.

come-nc committed:
chore: Turn method description into phpdoc

Co-authored-by: Ferdinand Thiessen <opensource@fthiessen.de>
Signed-off-by: Côme Chilliet <91878298+come-nc@users.noreply.github.com>

come-nc committed:
chore: Add comment to make code clearer

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

solracsf commented on lib/private/Share20/Manager.php:
```diff
@@ -1010,6 +1012,81 @@ protected function deleteChildren(IShare $share) {
+
+		foreach ($reshareRecords as $child) {
+			try {
+				/* Check if the share is still valid (means the resharer still has access to the file through another mean */
```
```suggestion
				/* Check if the share is still valid (means the resharer still has access to the file through another mean) */
```

come-nc commented:
Rebased and fixed code style.

come-nc committed:
fix(tests): Fix share tests to test new reshare promotion system

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
fix(tests): Revert changes to tests now that reshares are not deleted but promoted

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc commented on lib/private/Share20/Manager.php:
```diff
@@ -1039,6 +1041,81 @@ protected function deleteChildren(IShare $share) {
+			}
+
+			if ($node instanceof Folder) {
+				$sharesInFolder = $this->getSharesInFolder($userId, $node, false);
```
Shallow was removed in https://github.com/nextcloud/server/pull/44460

So, indeed, no idea how to get all shares in a folder subtree apart from recursing which would kill performance.
@icewind1991 Is there a trick we can use?

icewind1991 replied:
List all shares for a user and then filter them by path.

It's far from great but should be better than recursing into folders

come-nc committed:
fix: Fix promotion of reshares from subsubfolders

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc committed:
fix: Use getRelativePath method to check if node is inside folder

Signed-off-by: Côme Chilliet <come.chilliet@nextcloud.com>

come-nc commented:
I’m not sure if cypress failures can be related or not. Should be double check as sharing sidebar is on the failure screenshot.

susnux commented:
> I’m not sure if cypress failures can be related or not. Should be double check as sharing sidebar is on the failure screenshot.

Never saw that failing but is is 99,99% unrelated because what is tested is purely frontend behavior and unrelated to your changes

luka-nextcloud commented:
@come-nc I think need to create backport PRs for 28, 29, 30.

come-nc commented:
/backport to stable30

come-nc commented:
/backport to stable29

come-nc commented:
/backport to stable28
